### PR TITLE
feat(mix): extract `lockedVersion`

### PR DESCRIPTION
**Pull request by pdm-jd**
## Changes

Preliminary changes to the mix manager to support the `update-lockfile` versioning strategy. A separate PR for the manager and versioning portions was requested by @viceice 

Parent PR: https://github.com/renovatebot/renovate/pull/27915

## Context

- https://github.com/renovatebot/renovate/issues/13618

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

The changes only work by running with the versioning strategy changes in the parent PR as well.

### CODE REVIEW

**Comment by pdm-jd** [comment]
@rarkins I saw you just ran CI on this, thanks! The [`codecov` job failed](https://github.com/renovatebot/renovate/actions/runs/8283158098/job/22666451244) because it couldn't upload. Would it be possible to re-run?

File: lib/modules/manager/mix/extract.ts [hunk 1]
```diff
@@ -64,17 +67,35 @@ export async function extractPackageFile(
           }
         }
 
-        deps.push(dep);
+        deps.set(app, dep);
+        logger.info(`setting ${app} to ${JSON.stringify(dep)}`);
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:71]:
```suggestion
        logger.trace({dep}, `setting ${app}`);
```

File: lib/modules/manager/mix/extract.ts [hunk 2]
```diff
@@ -64,17 +67,35 @@ export async function extractPackageFile(
...
       }
     }
   }
-  const res: PackageFileContent = { deps };
   const lockFileName =
     (await findLocalSiblingOrParent(packageFile, 'mix.lock')) ?? 'mix.lock';
-  // istanbul ignore if
-  if (await localPathExists(lockFileName)) {
-    res.lockFiles = [lockFileName];
+  const lockFileContent = await readLocalFile(lockFileName);
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:78]:
```suggestion
  const lockFileContent = await readLocalFile(lockFileName, 'utf8');
```

parse as string

File: lib/modules/manager/mix/extract.ts [hunk 3]
```diff
@@ -64,17 +67,35 @@ export async function extractPackageFile(
...
+
+  if (lockFileContent) {
+    const lockedVersions = lockFileContent
+      .toString()
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:82]:
```suggestion
```

File: lib/modules/manager/mix/extract.ts [hunk 4]
```diff
@@ -64,17 +67,35 @@ export async function extractPackageFile(
...
+      .split(newlineRegex)
+      .slice(1, -1)
+      .reduce<Map<string, string>>((lockedVersions, line) => {
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:85]: use for-of loop instead

File: lib/util/pretty-time.spec.ts [hunk 5]
```diff
@@ -47,7 +47,7 @@ describe('util/pretty-time', () => {
   });
 
   describe('satisfiesDateRange()', () => {
-    const t0 = DateTime.fromISO('2023-07-07T12:00:00');
+    const t0 = DateTime.fromISO('2023-07-07T12:00:00Z');
```
Comment (viceice) [inline lib/util/pretty-time.spec.ts:50]: ?? revert unrelated changes

Comment (pdm-jd) [reply lib/util/pretty-time.spec.ts:50]: This test was failing locally, will revert.

File: lib/modules/manager/mix/extract.ts [hunk 6]
```diff
@@ -64,17 +67,35 @@ export async function extractPackageFile(
...
+      }
+    }
+
+    for (const [app, dep] of deps.entries()) {
+      dep.lockedVersion = lockedVersions.get(app);
+      logger.debug(`Found ${dep.lockedVersion} for ${app}`);
+    }
   }
-  return res;
+  return {
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:97]: return null when no deps found.

File: lib/modules/manager/mix/extract.ts [hunk 7]
```diff
@@ -64,17 +67,35 @@ export async function extractPackageFile(
...
+    for (const line of lockFileLines) {
+      const groups = lockedVersionRegExp.exec(line)?.groups;
+      if (groups?.app && groups?.lockedVersion) {
+        lockedVersions.set(groups.app, groups.lockedVersion);
+      }
+    }
+
+    for (const [app, dep] of deps.entries()) {
+      dep.lockedVersion = lockedVersions.get(app);
+      logger.debug(`Found ${dep.lockedVersion} for ${app}`);
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:94]:
```suggestion
      logger.trace(`Found ${dep.lockedVersion} for ${app}`);
```

File: lib/modules/manager/mix/extract.spec.ts [hunk 8]
```diff
@@ -25,8 +25,8 @@ describe('modules/manager/mix/extract', () => {
         {
           currentValue: '>2.1.0 or <=3.0.0',
           datasource: 'hex',
-          depName: 'ecto',
-          packageName: 'ecto',
+          depName: 'foo_bar',
+          packageName: 'foo_bar',
```
Comment (viceice) [inline lib/modules/manager/mix/extract.spec.ts:29]: why dep names are changed?

File: lib/modules/manager/mix/extract.ts [hunk 9]
```diff
@@ -64,17 +67,39 @@ export async function extractPackageFile(
...
+    for (const line of lockFileLines) {
+      const groups = lockedVersionRegExp.exec(line)?.groups;
+      if (groups?.app && groups?.lockedVersion) {
+        lockedVersions.set(groups.app, groups.lockedVersion);
+      }
+    }
+
+    for (const [app, dep] of deps.entries()) {
+      dep.lockedVersion = lockedVersions.get(app);
+      logger.trace(`Found ${dep.lockedVersion} for ${app}`);
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:94]: wrong, check if the locked version has this entry.

**Review by viceice — COMMENTED** [review]
see comments

File: lib/modules/manager/mix/extract.spec.ts (hunk 8 above)
Comment (pdm-jd) [reply lib/modules/manager/mix/extract.spec.ts:29]: I updated the dependencies in the `mix.exs` [fixture](https://github.com/renovatebot/renovate/pull/27924/files#diff-2d1a0f690965c612cc733c624ead42b46a906f2f2d889d6fad624721b8c74f8c) to be a valid. The previous version had two references to the `ecto` package, one to the Hex version and one to a Github version, which would fail to resolve in a real project.

File: lib/modules/manager/mix/extract.ts (hunk 9 above)
Comment (pdm-jd) [reply lib/modules/manager/mix/extract.ts:94]: Wouldn't have the same end result when checking for inclusion in `lockedVersions`?

`lockedVersions.get(app)` will return a string or undefined, both of which are valid types of `PackageDependency.lockedVersion`. The only difference is that the key will exist or not exist in the `dep` object.

The trace would either show that no locked version was found, "Found undefined for some_app", or the locked version, "Found 3.2.1 for some_app".

If you would like me to check, could you clarify where the inclusion check should go? Around both lines or just the trace?

File: lib/modules/manager/mix/extract.ts [hunk 10]
```diff
@@ -64,17 +67,39 @@ export async function extractPackageFile(
           }
         }
 
-        deps.push(dep);
+        deps.set(app, dep);
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:70]: can there be duplicates? if technically possible then log a warning

File: lib/modules/manager/mix/extract.ts [hunk 11]
```diff
@@ -64,17 +67,39 @@ export async function extractPackageFile(
...
     (await findLocalSiblingOrParent(packageFile, 'mix.lock')) ?? 'mix.lock';
-  // istanbul ignore if
-  if (await localPathExists(lockFileName)) {
-    res.lockFiles = [lockFileName];
+  const lockFileContent = await readLocalFile(lockFileName, 'utf8');
+
+  if (lockFileContent) {
+    const lockFileLines = lockFileContent.split(newlineRegex).slice(1, -1);
+
+    const lockedVersions: Map<string, string> = new Map();
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:83]:
```suggestion
    const lockedVersions = new Map<string, string>();
```
simplify 😉

File: lib/modules/manager/mix/extract.ts (hunk 9 above)
Comment (viceice) [reply lib/modules/manager/mix/extract.ts:94]: the message seems to be a little misleading/ confusing then 🤔

File: lib/modules/manager/mix/extract.ts (hunk 10 above)
Comment (pdm-jd) [reply lib/modules/manager/mix/extract.ts:70]: No, there cannot be multiple locked versions. It's one of the issues with the previous implementation, and why I had to change the `mix.exs` fixture to remove duplicates.

File: lib/modules/manager/mix/extract.ts (hunk 9 above)
Comment (pdm-jd) [reply lib/modules/manager/mix/extract.ts]: Updated to only print if a version was found.

File: lib/modules/manager/mix/extract.ts [hunk 12]
```diff
@@ -64,17 +67,42 @@ export async function extractPackageFile(
...
+  if (lockFileContent) {
+    const lockFileLines = lockFileContent.split(newlineRegex).slice(1, -1);
+
+    const lockedVersions = new Map<string, string>();
+
+    for (const line of lockFileLines) {
+      const groups = lockedVersionRegExp.exec(line)?.groups;
+      if (groups?.app && groups?.lockedVersion) {
+        lockedVersions.set(groups.app, groups.lockedVersion);
+      }
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:89]:
```suggestion
      const groups = lockedVersionRegExp.exec(line)?.groups;
      if (groups?.app && groups?.lockedVersion) {
        const dep = deps.get(groups.app);
        if (!dep) {
          logger.trace(`Missing dep ${groups.app} for locked version ${lockedVersion}`);
          continue;
        }
        dep.lockedVersion = groups.lockedVersion
        logger.trace(`Found ${lockedVersion} for ${groups.app}`);
      }
```

File: lib/modules/manager/mix/extract.ts [hunk 13]
```diff
@@ -64,17 +67,42 @@ export async function extractPackageFile(
...
+    }
+
+    for (const [app, dep] of deps.entries()) {
+      const lockedVersion = lockedVersions.get(app);
+      if (lockedVersion) {
+        dep.lockedVersion = lockedVersion;
+        logger.trace(`Found ${lockedVersion} for ${app}`);
+      }
+    }
+  }
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:99]:
```suggestion
```

File: lib/modules/manager/mix/extract.ts [hunk 14]
```diff
@@ -64,17 +67,42 @@ export async function extractPackageFile(
...
-  // istanbul ignore if
-  if (await localPathExists(lockFileName)) {
-    res.lockFiles = [lockFileName];
+  const lockFileContent = await readLocalFile(lockFileName, 'utf8');
+
+  if (lockFileContent) {
+    const lockFileLines = lockFileContent.split(newlineRegex).slice(1, -1);
+
+    const lockedVersions = new Map<string, string>();
+
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:84]:
```suggestion
```

the following changes saves us an iteration (perf)

Comment (pdm-jd) [reply lib/modules/manager/mix/extract.ts:84]: I see that these changes save one loop but I think it might make it harder to read the code and they do not preserve the same semantics as the original code. I would prefer to keep the code the way it is, but up to you. I will need to change it to preserve semantics.

1. Your change logs a trace if the dep isn't found in the lockfile. This will happen often because the lockfile will most likely have many more lines than the dependencies. In a large project I work on there are ~90 dependencies and 180 entries in the lockfile. Another project is 25 dependencies and 100 entries in the lockfile.
2.  In my opinion, the performance benefits are very very small. is making this change which will most likely make the code harder to read and harder to test worth saving a loop over 100 items?

Comment (viceice) [reply lib/modules/manager/mix/extract.ts:84]: 1. i dropped the log, we don't need it
2. we save a lot when you accumlate multiple renovate runs, so every second to save matters.

Comment (pdm-jd) [reply lib/modules/manager/mix/extract.ts]: Makes sense, thanks. Added your changes and ensured 100% coverage.

File: lib/modules/manager/mix/extract.ts [hunk 15]
```diff
@@ -64,17 +67,37 @@ export async function extractPackageFile(
...
-        deps.push(dep);
+        deps.set(app, dep);
+        logger.trace({ dep }, `setting ${app}`);
         depMatchGroups = depMatchRegExp.exec(depBuffer)?.groups;
       }
     }
   }
-  const res: PackageFileContent = { deps };
   const lockFileName =
     (await findLocalSiblingOrParent(packageFile, 'mix.lock')) ?? 'mix.lock';
```
Comment (viceice) [inline lib/modules/manager/mix/extract.ts:77]: are you sure it always exists? the read below can fail if file isn't found

Comment (pdm-jd) [reply lib/modules/manager/mix/extract.ts:77]: It should always exist. It always exists if you've fetched dependencies. I don't see the use case of running renovate on a repository where you haven't fetched dependencies before.

`readLocalFile/2` catches exceptions and returns `null` [ref](https://github.com/renovatebot/renovate/blob/main/lib/util/fs/index.ts#L26-L44), so this shouldn't cause an exception since we check the truthiness of `lockFileContent` before iterating through the content.

**Review by viceice — APPROVED** [review]
